### PR TITLE
Geo mle integration

### PR DESCRIPTION
**Pull request by Malinon**

### CODE REVIEW

File: skdim/id_flex/_GeoMLE.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            raise ValueError("Number of bootstrap sets has to  be positive")
+        if self.max_degree <= 0:
+            raise ValueError("Degree of interpolation polynomial has to be positive")
+        if self.nbhd_type not in ['knn']:
+            raise ValueError('Neighbourhood type should be knn')
+        
+        k2 = self.n_neighbors + self.average_steps - 1
+        self.dimension_pw_ = self.geomle(X, self.n_neighbors, k2)
+    
+    def geomle(self, X, k1, k2):
```
Comment (johnharveymath) [inline skdim/id_flex/_GeoMLE.py:58]: private?

**Review by johnharveymath — COMMENTED** [review]
I'll take another look once it's passing all tests, but it seems good.

File: skdim/id_flex/_GeoMLE.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    
+        Input parameters:
+        X            - data
+        k1           - minimal number of nearest neighbours
+        k2           - maximal number of nearest neighbours
+        max_degree   - maximal degree of polynomial regression (Default = 2)
+        metric       - metric for the distance calculation (Default = 'euclidean')
+    
+        Returns: 
+        array of shape (nb_iter1,) of regression dimensionality estimation for k = k1..k2 averaged over bootstrap samples
```
Comment (kmyim) [inline skdim/id_flex/_GeoMLE.py:70]: what is nb_iter1 ?

File: skdim/id_flex/_GeoMLE.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        """
+        if self.metric == 'precomputed':
+            dist = X
+            DIM_SPACE = None
+        else:
+            DIM_SPACE = X.shape[1]
+            dist = pairwise_distances(X, X, metric=self.metric)
+        NUM_OF_POINTS = X.shape[0]
+
+        result = []
```
Comment (kmyim) [inline skdim/id_flex/_GeoMLE.py:80]: lists 'results' and 'reg' can be removed

File: skdim/id_flex/_GeoMLE.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            bootstrap_ids = GeoMle.__gen_bootstrap_ids(X)
+            # Calculate distances to nearest neighbours in bootsrap dataset
+            for x_id in range(X.shape[0]):
+                bootstrap_distances_nn = np.sort(GeoMle._get_nearest_distances_in_bootstrap(x_id, dist, bootstrap_ids, k2))
+                # update average distances
+                avg_distances[x_id, :] += bootstrap_distances_nn[k1 - 1:]
+                for k_iter in range(self.average_steps):
+                    # Calculate MLE for bootstrap sample
+                    mles[x_id, j, k_iter] = GeoMle._calc_mle(bootstrap_distances_nn[:k_iter + k1])
+                    #print("Estims", mles[x_id, j, k_iter])
```
Comment (kmyim) [inline skdim/id_flex/_GeoMLE.py:95]: commented out print can be removed

File: skdim/tests/flex/test_FlexNbhdEstimator.py [hunk 5]
```diff
@@ -65,22 +65,28 @@ def test_fit_pw_computes_pw_dim(estim_class, data):
...
 def test_fit_pw_single_job_equal_to_fit_pw(estim_class, data):
     estimator_single = estim_class()
-    estimator_single._fit_pw(data)
-    estimator_multi = estim_class(n_jobs=4)
-    estimator_multi._fit_pw(data)
+    np.random.seed(0)
+    estimator_single.fit_pw(data)
+    estimator_multi = estim_class()
+    np.random.seed(0)
+    estimator_multi.fit_pw(data, n_jobs=4)
```
Comment (kmyim) [inline skdim/tests/flex/test_FlexNbhdEstimator.py:82]: move n_jobs param to `estimator_multi = estim_class(n_jobs = 4)`

File: skdim/tests/flex/geoMle_test.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
+import pytest
+import numpy as np
+import skdim.id_flex
+import geomle as gm
```
[the remark this answers is not in the document]
Comment (kmyim) [reply skdim/tests/flex/geoMle_test.py:4]: should be fine if geomle package installed https://github.com/stat-ml/GeoMLE

File: skdim/id_flex/_GeoMLE.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+class GeoMle(FlexNbhdEstimator):
+    def __init__(self, average_steps = 2, bootstrap_num = 20, alpha = 5e-3, interpolation_degree = 2,
+        metric="euclidean",
+        comb="mean",
+        smooth=False,
+        n_jobs=1,
+        n_neighbors=5):
+        super().__init__(
+            pw_dim=True,
+            nbhd_type="knn",
```
[the remark this answers is not in the document]
Comment (Malinon) [reply skdim/id_flex/_GeoMLE.py:16]: It is fixed in newest commit. Now tests accept also TypeError

**Review by kmyim — APPROVED** [review]

**Pull request merged**